bgpv1: component test framework

harsimran-pabla opened the pull request:
Introducing bgp control plane component tests. This change introduces test harness and basic BGP tests related to podCIDR advertisements, lb service advertisements and neighbor events.

## Motivation of this change

### Increase test coverage around boundaries of subcomponents

- For eg, we do not test complete flow of change from getting Kubernetes event and then validating that it is reflected in the peering router.
- Unit tests today exists for
-- Controller : This is limited to that policy change calls correct reconciler methods.
-- Reconcilers : Tests validate the diffs are generated correctly and we are inspecting gobgp directly for validation. Direct inspection of gobgp might not be suitable here in long term, if another underlying routing agent is used.

### Validate and make BGP Peering part of CI testing.

- This is mainly for neighbour configuration : like timers, graceful restart, authentication etc. And later on for more advanced BGP as well ( Routing policies and filters )
- During new feature development, we have to validate the expected behaviour and add tests for it. It is currently not possible with unit tests, as we only configure local gobgp and we have to validate the behaviour separately via container labs to test actual peering behaviour. Which is again not coded anywhere and we do not repeat those manual tests as we continue to modify the code ( I would say this is the biggest reason ).
- Also having this tests as part of CI, will ensure that there are no regressions introduced to feature behavior as we make changes.

## Types of test cases to add

This component level test is mainly cover high level BGP features

So, what I see following tests need to be added

* Pod CIDR advertisements
* LB Service advertisements
* Neighbor behavior
-- Authentication
-- Timers
-- Graceful restart
* Some special test cases
-- Restarts
-- Any customer reported bugs and issues with order of operations.

YutaroHayakawa commented on pkg/bgpv1/test/adverts_test.go:
```diff
@@ -0,0 +1,360 @@
+	slimv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	"github.com/cilium/cilium/pkg/testutils"
+
+	"github.com/stretchr/testify/require"
```
Just a question: What's the motivation for introducing the new test framework?

harsimran-pabla replied:
So motivation is two fold, and goes hand in hand with current unit tests.
- Increase test coverage around boundaries of subcomponents
    - For eg, we do not test complete flow of change from getting Kubernetes event and then validating that it is reflected in the peering router.
    - Unit tests today exists for 
        - Controller : This is limited to that policy change calls correct reconciler methods.
        - Reconcilers : Tests validate the diffs are generated correctly and we are inspecting gobgp directly for validation. 
            - Direct inspection of gobgp might not be suitable here in long term, if another underlying routing agent is used.
- Validate and make BGP Peering part of CI testing.
    - This is mainly for neighbour configuration : like timers, graceful restart, authentication etc. And later on for more advanced BGP as well ( Routing policies and filters )
    - During new feature development, we have to validate the expected behaviour and add tests for it. It is currently not possible with unit tests, as we only configure local gobgp and we have to validate the behaviour separately via container labs - to test actual peering behaviour. Which is again not coded anywhere and we do not repeat those manual tests as we continue to modify the code ( I would say this is the biggest reason ). 
    - Also having this tests as part of CI, will ensure that there are no regressions introduced to feature behavior as we make changes.

YutaroHayakawa replied:
Ah, I was just asking the reason for introducing `github.com/stretchr/testify/require` for curiosity 😓 

But, thanks for the clarification. I think it would be nice to have these contexts in the commit message.

harsimran-pabla replied:
Ah, so there are two packages in testify, assert and require. From what I understand, on any failure, assert will continue the test run and execute as many sub-tests as possible and eventually fail the test case. 

Whereas, require will fail the test case on failing step and not proceed further. I personally prefer this approach, so time and resources are not spent on test case which is failing anyways. Also, it provides quicker feedback while doing development.

harsimran-pabla replied:
Also, testify package provides interesting validation patterns. Like comparing sets, values, error/noerror etc. Which makes test more readable.

YutaroHayakawa replied:
Alright, that makes sense. I don't really have any preference around the testing framework, so no objection 👍

YutaroHayakawa commented on pkg/bgpv1/test/adverts_test.go:
```diff
@@ -0,0 +1,360 @@
+	maxTestDuration = 15 * time.Second
+)
+
+// Test_PodCIDRAdvert validates pod IPv4 subnet is advertised, withdrawn and modified on node addresses change.
```
Nit: IPv4 => IPv4/IPv6? We test both of them.

YutaroHayakawa commented on pkg/bgpv1/test/adverts_test.go:
```diff
@@ -0,0 +1,360 @@
+					prefix:      "aaaa::",
+					prefixLen:   64,
+					isWithdrawn: false,
+				},
+			},
+		},
+		{
+			description: "update pod CIDRs",
+			podCIDRs: []string{
+				"10.2.0.0/16",
+				"bbbb::/64",
+			},
+			expectedRouteEvents: []routeEvent{
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "10.1.0.0",
+					prefixLen:   16,
+					isWithdrawn: true,
+				},
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "10.2.0.0",
+					prefixLen:   16,
+					isWithdrawn: false,
+				},
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "aaaa::",
+					prefixLen:   64,
+					isWithdrawn: true,
+				},
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "bbbb::",
+					prefixLen:   64,
+					isWithdrawn: false,
+				},
+			},
+		},
+	}
```
How can we choose the test case covered in this component test? For example, we can test this re-advertise case with the reconciler's unit test. How did you choose this case?

harsimran-pabla replied:
This component level test is mainly BDD ( behavior driven development ), where high level features are tested. 

So, what I see following tests to be added 
- Pod CIDR advertisements
- LB Service advertisements
- Neighbor behavior
    - Authentication
    - Timers
    - Graceful restart

Some special test cases
- Restarts 
- Any customer reported bugs and issues with order of operations. 

I think overall, we need to add unit tests in reconciler, controller and gobgp shim layer while doing the development. Whereas, under test/ directory, we should have high level tests which show case BGP features.

harsimran-pabla replied:
I think reconciler test is doing bit beyond it's role, it should be basically testing just diffs which is the logical processing reconciler is doing. I think validating gobgp internals from reconciler is leaking underlying abstraction of gobgp into generic reconciler code.

YutaroHayakawa replied:
Thanks for your clarification. The BDD description makes sense to me 👍 We should also cover those descriptions in the commit message.

harsimran-pabla replied:
Strictly speaking, BDD would require some sort of formal language for defining test cases - like gherkin (godog). 

Using go tests provide more flexibility, but could be difficult to read if steps are not defined clearly. I will add this in the description.

YutaroHayakawa commented on pkg/bgpv1/test/adverts_test.go:
```diff
@@ -0,0 +1,360 @@
+		description         string
+		srvName             string
+		ingressIP           string
+		oper                string // add or update
```
Nit: What's `oper` stands for? Maybe "Operation"? Then it's common to use `op` I think.

YutaroHayakawa commented on pkg/bgpv1/test/links.go:
```diff
@@ -0,0 +1,119 @@
+		// link used by gobgp instance 2
+		instance2Link: {
+			ipv4: netip.MustParsePrefix("172.16.100.3/32"),
+			ipv6: netip.MustParsePrefix("a::3/128"),
```
These IP addresses may overlap with other networks like Docker, VirtualBox, etc. So, it's better to run tests inside the dedicated network namespace and create dummy interfaces there so that we can use whatever address we want.

harsimran-pabla replied:
This turned out to be more tricky, not exactly sure on how to fix it.

Setting netns from within go test does not work well, as new go routines are not part of same netns. 
https://github.com/containernetworking/plugins/blob/main/pkg/ns/README.md 

Links and addrs are created inside a new namespace, but gobgp starts listening in parent ns.

harsimran-pabla replied:
I wonder, once privilege tests are moved from ginkgo to github actions, we can wrap this via shell in github workflow ?

```
ip netns add bgp-test
ip netns bgp-test exec go test <bgp tests>
ip netns delete bgp-test
```

YutaroHayakawa replied:
Ugh, alright. That's indeed tricky... I could come up with three solutions, but I'm not sure how hard those are...

1. Spin up multiple GoBGP instances locally with different port numbers. We first need to support changing the peer's port number to implement this.
2. Run peer GoBGP daemons separately in the different processes in different netns.
3. Make IP addresses configurable

harsimran-pabla replied:
> Spin up multiple GoBGP instances locally with different port numbers. We first need to support changing the peer's port number to implement this.

Currently, that is happening. Peering gobgp instances are started with different port number. Since they are the ones connecting to BGP CP gobgp, it works fine. 

> Run peer GoBGP daemons separately in the different processes in different netns.

I'd like to have control over gobgp instances from within test driver - so we can inspect its internal state. I have a feeling this is going to be complicated if we start gobgp from outside the scope of test driver. 

> Make IP addresses configurable

I am not sure what you mean by this one ? Like pass IP in test environment  ?

Maybe short term, I can make IPs overridable via env settings. If that is not enough, then look at creating dedicated github workflow to create netns before running this test case.

YutaroHayakawa replied:
> Like pass IP in test environment ?

Yep, that's right. Seems like you already did this in the latest commit. We've also talked out-of-band and agreed to go with a configurable IP address in the short term.

YutaroHayakawa commented on pkg/bgpv1/gobgp/logger.go:
When you brush up on the commits, please consider moving this refactoring into a different commit.

harsimran-pabla committed:
bgpv1: pass parameters to gobgp logger for reusability

Pass log parameters to gobgp logger instead of hardcoding them in log
methods. This is done so we can reuse the logger and pass on different
log fields.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgpv1: Introduction of bgp component tests

Introducing bgp control plane component tests. This change introduces
test harness and basic BGP tests related to podCIDR advertisements, lb
service advertisements and neighbor events.

Motivation of this change is to increase test coverage around boundaries
of BGP subsystem and also to validate actual peering state at high
level for primary BGP features.

Types of test cases to add here would range from exposed BGP features
like route advertisements and various neighbor configuration settings.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

YutaroHayakawa reviewed (approved):
Now looks good to me 👍 

One non-blocking comment: I'd prefer putting all context in the PR description into the commit message. So that we don't have to jump to GitHub every time.
